core/models/network.go

```diff
@@ -15,8 +21,63 @@ type NetworkInterface struct {
 	GuestMAC string `json:"guest_mac,omitempty" validate:"omitempty,mac"`
 	// Type is the type of host network interface type to create to use by the guest.
 	Type IfaceType `json:"type" validate:"oneof=tap macvtap unsupported"`
-	// Address is an optional IP address to assign to this interface. If not supplied then DHCP will be used.
-	Address string `json:"address,omitempty" validate:"omitempty,cidr"`
+	// StaticAddress is an optional static IP address to assign to this interface.
+	// If not supplied then DHCP will be used.
+	StaticAddress *StaticAddress `json:"staticAddress,omitempty"`
+}
+
+// StaticAddress specifies a static IP address configuration.
+type StaticAddress struct {
+	// Address is the static IP address (IPv4 or IPv6) to assign to this interface.
+	// Must be CIDR notation.
+	Address IPAddressCIDR `json:"address" validate:"cidr"`
+	// Gateway is used to optionally set the default gateway for IPv4 or IPv6.
+	Gateway *IPAddressCIDR `json:"gateway,omitempty" validate:"omitempty,cidr"`
+	// Nameservers allows you to optionally specify nameservers for the interface.
+	Nameservers []string `json:"nameservers" validate:"omitempty,dive,ip"`
+}
+
+// IPAddressCIDR represents a IPv4/IPv6 address in CIDR notation.
+type IPAddressCIDR string
+
+func (i IPAddressCIDR) IsIPv4() (bool, error) {
+	ip, _, err := net.ParseCIDR(string(i))
+	if err != nil {
+		return false, fmt.Errorf("parsing %s as cidr: %w", i, err)
+	}
+
+	// 0:0:0:0:0:ffff:0101:0101 is the ipv6 representation of 1.1.1.1, the net
+	// package can parse and convert to IPv4. Based on Slack messages, we want a
+	// strict validation, they have to spcify an IPv4 CIDR block.
+	containsDot := strings.Contains(string(i), ".")
+
+	return ip.To4() != nil && containsDot, nil
+}
+
+func (i IPAddressCIDR) IsIPv6() (bool, error) {
```

Comment:
am I blind? I don't think this is actually being used anywhere other than the tests

Reply:
I think it was added to be there "in case we want to use for validation"

Reply:
hmmm another of those rainy day pieces?

Reply:
let's kill it i think, wont be too hard to copy-pasta the above if we actually end up needing it

Reply:
I would still let it live there, it makes sense if we have an IsIPv4 method, then we have an IsIPv6 method too and the function itself captures the intended logic (strict ipv6 format check as all ipv4 addresses are valid ipv6 address).

Reply:
So in the future if it's not there, there is a chance someone says "yeah 192.168.12.8 is a valid ipv6 address" and we want to check if it's a full ipv6 format or not.

Reply:
but it's dead code tho, right? we can comment on the isipv4 method and say "this doesn't necessarily mean that it is ipv6". there is no need to use dead code for a comment when we could just... have a comment.

Reply:
Removed that function.

---

I'm more on the side of "implement known logic if they are in pair with other functions", like if we haven Either struct that has a Left and a Right value, we set them as private fields, but because we don't use `Left()` yet in the code, it does not mean we should implement only the `Right()` and say we will add Left when we want to use it somewhere outside and not only struct-internal. In this case, it makes total sens to me if a we have a package that has `IsIPv4` method, then it has an `IsIPv6` method too when I want to use it somewhere else, and I'll not get mad because I have to double check if my editor is not able to autocomplete it or it's not implemented at all and go there and implement.

Reply:
I do get that, and I also conflictingly have the same opinion.

The problem is that I have two opposing rules: all code should be in use, otherwise it has no reason for existence; and exactly what you explained above.

For me the first rule wins because I can't justify "I like symmetry" for a non-library package 😭

Reply:
> non-library package
that's a `core/model` package, it's like an internal lib package